src/lib/mina/local-blockchain.ts

```diff
@@ -0,0 +1,389 @@
+      // events are identified and associated with a publicKey and tokenId
+      txn.transaction.accountUpdates.forEach((p, i) => {
+        let pJson = zkappCommandJson.accountUpdates[i];
+        let addr = pJson.body.publicKey;
+        let tokenId = pJson.body.tokenId;
+        events[addr] ??= {};
+        if (p.body.events.data.length > 0) {
+          events[addr][tokenId] ??= [];
+          let updatedEvents = p.body.events.data.map((data) => {
+            return {
+              data,
+              transactionInfo: {
+                transactionHash: '',
+                transactionStatus: '',
+                transactionMemo: '',
+              },
+            };
+          });
+          events[addr][tokenId].push({
+            events: updatedEvents,
+            blockHeight: networkState.blockchainLength,
+            globalSlot: networkState.globalSlotSinceGenesis,
+            // The following fields are fetched from the Mina network. For now, we mock these values out
+            // since networkState does not contain these fields.
+            blockHash: '',
+            parentBlockHash: '',
+            chainStatus: '',
+          });
+        }
+
+        // actions/sequencing events
+
+        // most recent action state
+        let storedActions = actions[addr]?.[tokenId];
+        let latestActionState_ =
+          storedActions?.[storedActions.length - 1]?.hash;
+        // if there exists no hash, this means we initialize our latest hash with the empty state
+        let latestActionState =
+          latestActionState_ !== undefined
+            ? Field(latestActionState_)
+            : Actions.emptyActionState();
+
+        actions[addr] ??= {};
+        if (p.body.actions.data.length > 0) {
+          let newActionState = Actions.updateSequenceState(
+            latestActionState,
+            p.body.actions.hash
+          );
+          actions[addr][tokenId] ??= [];
+          actions[addr][tokenId].push({
+            actions: pJson.body.actions,
+            hash: newActionState.toString(),
+          });
+        }
+      });
+
+      const hash = Test.transactionHash.hashZkAppCommand(txn.toJSON());
+      const pendingTransaction: Omit<
+        PendingTransaction,
+        'wait' | 'waitOrThrowIfError'
+      > = {
+        isSuccess,
+        errors,
+        transaction: txn.transaction,
+        toJSON: txn.toJSON,
+        toPretty: txn.toPretty,
+        hash: (): string => {
+          return hash;
+        },
+      };
+
+      const wait = async (_options?: {
+        maxAttempts?: number;
+        interval?: number;
+      }) => {
+        return createIncludedOrRejectedTransaction(
+          pendingTransaction,
+          pendingTransaction.errors
+        );
+      };
+
+      const waitOrThrowIfError = async (_options?: {
+        maxAttempts?: number;
+        interval?: number;
+      }) => {
+        return createIncludedOrRejectedTransaction(
```

Comment:
for consistency, I think this should behave like the `Network` implementation of `waitOrThrowIfError()`, and throw an error if the pending transaction already failed

Reply:
fixed!